src/mgr/DaemonServer.cc

```diff
@@ -1766,6 +1766,188 @@ void DaemonServer::send_report()
   monc->send_mon_message(m);
 }
 
+void DaemonServer::adjust_pgs()
+{
+  dout(20) << dendl;
+  unsigned max = std::max<int64_t>(1, g_conf->mon_osd_max_creating_pgs);
+
+  map<string,unsigned> pg_num_to_set;
+  map<string,unsigned> pgp_num_to_set;
+  cluster_state.with_pgmap([&](const PGMap& pg_map) {
+      unsigned creating_or_unknown = 0;
+      for (auto& i : pg_map.num_pg_by_state) {
+	if ((i.first & (PG_STATE_CREATING)) ||
+	    i.first == 0) {
+	  creating_or_unknown += i.second;
+	}
+      }
+      unsigned left = max;
+      if (creating_or_unknown >= max) {
+	return;
+      }
+      dout(10) << "creating_or_unknown " << creating_or_unknown
+	       << " max_creating " << max << dendl;
+      cluster_state.with_osdmap([&](const OSDMap& osdmap) {
+	  if (pg_map.last_osdmap_epoch != osdmap.get_epoch()) {
+	    // do nothing if maps aren't in sync
+	    dout(10) << "last_osdmap_epoch " << pg_map.last_osdmap_epoch
+		     << " osdmap " << osdmap.get_epoch() << dendl;
+	    //return;
+	  }
+	  for (auto& i : osdmap.get_pools()) {
+	    const pg_pool_t& p = i.second;
+
+	    // adjust pg_num?
+	    if (p.get_pg_num_target() != p.get_pg_num()) {
+	      dout(20) << "pool " << i.first
+		       << " pg_num " << p.get_pg_num()
+		       << " target " << p.get_pg_num_target()
+		       << dendl;
+	      if (p.has_flag(pg_pool_t::FLAG_CREATING)) {
+		dout(10) << "pool " << i.first
+			 << " target " << p.get_pg_num_target()
+			 << " pg_num " << p.get_pg_num()
+			 << " - still creating initial pgs"
+			 << dendl;
+	      } else if (p.get_pg_num() != p.get_pg_num_pending()) {
+		dout(10) << "pool " << i.first
+			 << " target " << p.get_pg_num_target()
+			 << " pg_num " << p.get_pg_num()
+			 << " - pg_num_pending != pg_num, waiting"
+			 << dendl;
+		// FIXME: we might consider allowing pg_num increases without
+		// waiting for the previously planned merge to complete.
+	      } else if (p.get_pg_num_target() < p.get_pg_num()) {
+		// pg_num decrease (merge)
```

Comment:
should we guard this with osd_release >= MIMIC to avoid spamming the monitors with commands that get EINVAL?

Reply:
the mon command that allows you to lower pg_num_target is already guarding for require_osd_release, so we'd never reach this point with luminous osds